File: apps/cargo-tracker/src/CargoTracker.Web/Booking/Interfaces/BookingForm.cs
using System.ComponentModel.DataAnnotations;
using CargoTracker.Booking.Application.Internal.QueryServices;
using CargoTracker.Booking.Domain.Model;

namespace CargoTracker.Booking.Interfaces;

/// <summary>貨物予約登録フォーム（US04）。</summary>
public sealed class BookingForm
{
    [Required(ErrorMessage = "荷主 ID を選択してください")]
    [Display(Name = "荷主 ID")]
    public string ShipperId { get; set; } = string.Empty;

    [Required(ErrorMessage = "出発地（UN/LOCODE）を入力してください")]
    [StringLength(5, MinimumLength = 5, ErrorMessage = "UN/LOCODE は 5 文字です")]
    [Display(Name = "出発地（UN/LOCODE）")]
    public string OriginUnLocode { get; set; } = string.Empty;

    [Required(ErrorMessage = "仕向地（UN/LOCODE）を入力してください")]
    [StringLength(5, MinimumLength = 5, ErrorMessage = "UN/LOCODE は 5 文字です")]
    [Display(Name = "仕向地（UN/LOCODE）")]
    public string DestinationUnLocode { get; set; } = string.Empty;

    [Required(ErrorMessage = "希望着日を入力してください")]
    [DataType(DataType.Date)]
    [Display(Name = "希望着日")]
    public DateOnly ArrivalDeadline { get; set; }

    [Required]
    [Display(Name = "貨物種別")]
    public string CargoType { get; set; } = "General";

    [Range(0.001, 1_000_000, ErrorMessage = "重量は正の値を入力してください")]
    [Display(Name = "重量（kg）")]
    public decimal Weight { get; set; }

    [Display(Name = "長さ（cm）")]
    public decimal? DimensionLength { get; set; }

    [Display(Name = "幅（cm）")]
    public decimal? DimensionWidth { get; set; }

    [Display(Name = "高さ（cm）")]
    public decimal? DimensionHeight { get; set; }

    [Range(1, int.MaxValue, ErrorMessage = "個数は 1 以上で入力してください")]
    [Display(Name = "個数")]
    public int? Quantity { get; set; }

    [StringLength(500, ErrorMessage = "品名は 500 文字以内で入力してください")]
    [Display(Name = "品名")]
    public string? Description { get; set; }

    [Display(Name = "危険物クラス")]
    public string? HazardousClass { get; set; }

    [Display(Name = "UN 番号")]
    public string? UnNumber { get; set; }

    [Display(Name = "正式輸送品名")]
    public string? ProperShippingName { get; set; }

    [Display(Name = "最低温度")]
    public decimal? MinTemperature { get; set; }

    [Display(Name = "最高温度")]
    public decimal? MaxTemperature { get; set; }

    [Display(Name = "温度単位")]
    public TemperatureUnit? TemperatureUnit { get; set; }

    public IReadOnlyList<ShipperOption> Shippers { get; set; } = [];
}

File: apps/cargo-tracker/src/main/java/com/example/cargotracker/booking/interfaces/web/BookingForm.java
package com.example.cargotracker.booking.interfaces.web;

import jakarta.validation.constraints.Min;
import jakarta.validation.constraints.NotBlank;
import jakarta.validation.constraints.NotNull;
import jakarta.validation.constraints.Pattern;
import jakarta.validation.constraints.Size;
import java.math.BigDecimal;
import java.time.LocalDate;
import org.springframework.format.annotation.DateTimeFormat;

/**
 * 貨物予約登録フォーム（US04）。
 *
 * <p>画面の検証は案内であり、<strong>最後の砦はドメインと DB の制約</strong>である。
 * ただし案内と実際の受け入れ条件がずれていてはならないため、形式はドメインの
 * 値オブジェクトと揃える（UN/LOCODE 5 文字、重量は正の値、個数は 1 以上）。
 *
 * <p>出発地と目的地が同じ場合・到着期限が過去の場合の拒否は、
 * 項目単体では判定できないためドメインが行う。
 */
public class BookingForm {

    /** 荷主コード。荷主 ID ではなく業務コードで指定する（暗記を前提にしない）。 */
    @NotBlank(message = "荷主コードは必須です")
    @Pattern(regexp = "SHP-\\d{6}", message = "荷主コードは SHP-999999 形式です")
    private String shipperCode;

    @NotBlank(message = "出発地は必須です")
    @Pattern(regexp = "[A-Z]{2}[A-Z0-9]{3}", message = "出発地は UN/LOCODE（英大文字 5 文字）です")
    private String origin;

    @NotBlank(message = "目的地は必須です")
    @Pattern(regexp = "[A-Z]{2}[A-Z0-9]{3}", message = "目的地は UN/LOCODE（英大文字 5 文字）です")
    private String destination;

    @NotNull(message = "希望到着期限は必須です")
    @DateTimeFormat(iso = DateTimeFormat.ISO.DATE)
    private LocalDate arrivalDeadline;

    @NotBlank(message = "貨物種別は必須です")
    @Pattern(regexp = "GENERAL|HAZARDOUS|REFRIGERATED", message = "貨物種別が不正です")
    private String cargoType = "GENERAL";

    @NotNull(message = "重量は必須です")
    @jakarta.validation.constraints.DecimalMin(
            value = "0.001", message = "重量は 0 より大きい値です")
    @jakarta.validation.constraints.Digits(
            integer = 7, fraction = 3, message = "重量は小数第 3 位までです")
    private BigDecimal weight;

    @Min(value = 1, message = "個数は 1 以上です")
    private Integer quantity;

    private BigDecimal dimensionLength;
    private BigDecimal dimensionWidth;
    private BigDecimal dimensionHeight;

    @Size(max = 500, message = "品名は 500 文字までです")
    private String description;

    /**
     * 危険物クラス（US05。<strong>危険物のときだけ必須</strong>）。
     *
     * <p><strong>必須の判断はここでは行わない。</strong> 種別と申告の整合は
     * {@code CargoSpecification} が守る。ここに条件を書き写すと規則が 2 か所に散る
     * （US03 の契約情報と同じ扱い）。
     */
    private String hazardClass;

    /** UN 番号（US05）。 */
    private String unNumber;

    /** 正式輸送品名（US05）。 */
    private String properShippingName;

    /** 最低温度（US05。<strong>冷凍・冷蔵のときだけ必須</strong>）。 */
    private java.math.BigDecimal minTemperature;

    /** 最高温度（US05）。 */
    private java.math.BigDecimal maxTemperature;

    /** 温度の単位（{@code CELSIUS} / {@code FAHRENHEIT}）。 */
    private String temperatureUnit;

    public String getHazardClass() {
        return hazardClass;
    }

    public void setHazardClass(String hazardClass) {
        this.hazardClass = hazardClass;
    }

    public String getUnNumber() {
        return unNumber;
    }

    public void setUnNumber(String unNumber) {
        this.unNumber = unNumber;
    }

    public String getProperShippingName() {
        return properShippingName;
    }

    public void setProperShippingName(String properShippingName) {
        this.properShippingName = properShippingName;
    }

    public java.math.BigDecimal getMinTemperature() {
        return minTemperature;
    }

    public void setMinTemperature(java.math.BigDecimal minTemperature) {
        this.minTemperature = minTemperature;
    }

    public java.math.BigDecimal getMaxTemperature() {
        return maxTemperature;
    }

    public void setMaxTemperature(java.math.BigDecimal maxTemperature) {
        this.maxTemperature = maxTemperature;
    }

    public String getTemperatureUnit() {
        return temperatureUnit;
    }

    public void setTemperatureUnit(String temperatureUnit) {
        this.temperatureUnit = temperatureUnit;
    }

    public String getShipperCode() {
        return shipperCode;
    }

    public void setShipperCode(String shipperCode) {
        this.shipperCode = shipperCode;
    }

    public String getOrigin() {
        return origin;
    }

    public void setOrigin(String origin) {
        this.origin = origin;
    }

    public String getDestination() {
        return destination;
    }

    public void setDestination(String destination) {
        this.destination = destination;
    }

    public LocalDate getArrivalDeadline() {
        return arrivalDeadline;
    }

    public void setArrivalDeadline(LocalDate arrivalDeadline) {
        this.arrivalDeadline = arrivalDeadline;
    }

    public String getCargoType() {
        return cargoType;
    }

    public void setCargoType(String cargoType) {
        this.cargoType = cargoType;
    }

    public BigDecimal getWeight() {
        return weight;
    }

    public void setWeight(BigDecimal weight) {
        this.weight = weight;
    }

    public Integer getQuantity() {
        return quantity;
    }

    public void setQuantity(Integer quantity) {
        this.quantity = quantity;
    }

    public BigDecimal getDimensionLength() {
        return dimensionLength;
    }

    public void setDimensionLength(BigDecimal dimensionLength) {
        this.dimensionLength = dimensionLength;
    }

    public BigDecimal getDimensionWidth() {
        return dimensionWidth;
    }

    public void setDimensionWidth(BigDecimal dimensionWidth) {
        this.dimensionWidth = dimensionWidth;
    }

    public BigDecimal getDimensionHeight() {
        return dimensionHeight;
    }

    public void setDimensionHeight(BigDecimal dimensionHeight) {
        this.dimensionHeight = dimensionHeight;
    }

    public String getDescription() {
        return description;
    }

    public void setDescription(String description) {
        this.description = description;
    }
}
